sklearnex/cluster/k_means.py

```diff
@@ -1,25 +1,373 @@
+                "random_state": self.random_state,
+            }
+
+            self._onedal_estimator = onedal_KMeans(**onedal_params)
+
+        def _onedal_fit_supported(self, method_name, X, y=None, sample_weight=None):
+            assert method_name == "fit"
+
+            class_name = self.__class__.__name__
+            patching_status = PatchingConditionsChain(f"sklearn.cluster.{class_name}.fit")
+
+            sample_count = _num_samples(X)
+            self._algorithm = self.algorithm
+            supported_algs = ["auto", "full", "lloyd", "elkan"]
+            if self.algorithm == "elkan":
+                logging.getLogger("sklearnex").info(
+                    "oneDAL does not elkan, using lloyd algorithm instead."
+                )
+            correct_count = self.n_clusters < sample_count
+
+            is_data_supported = (
+                _is_csr(X) and daal_check_version((2024, "P", 700))
+            ) or not issparse(X)
+
+            _acceptable_sample_weights = True
+            if sample_weight is not None:
+                sample_weight = _check_sample_weight(
+                    sample_weight, X, dtype=X.dtype if hasattr(X, "dtype") else None
+                )
+                _acceptable_sample_weights = np.allclose(
+                    sample_weight, np.ones_like(sample_weight)
+                )
+
+            patching_status.and_conditions(
+                [
+                    (
+                        self.algorithm in supported_algs,
+                        "Only lloyd algorithm is supported, elkan is computed using lloyd",
+                    ),
+                    (correct_count, "n_clusters is smaller than number of samples"),
+                    (
+                        _acceptable_sample_weights,
+                        "oneDAL doesn't support sample_weight, either None or ones are acceptable",
+                    ),
+                    (
+                        is_data_supported,
+                        "Supported data formats: Dense, CSR (oneDAL version >= 2024.7.0).",
+                    ),
+                ]
+            )
+
+            return patching_status
+
+        def fit(self, X, y=None, sample_weight=None):
+            if sklearn_check_version("1.2"):
+                self._validate_params()
+
+            dispatch(
+                self,
+                "fit",
+                {
+                    "onedal": self.__class__._onedal_fit,
+                    "sklearn": sklearn_KMeans.fit,
+                },
+                X,
+                y,
+                sample_weight,
+            )
+
+            return self
+
+        def _onedal_fit(self, X, _, sample_weight, queue=None):
+            X = self._validate_data(
+                X,
+                accept_sparse="csr",
+                dtype=[np.float64, np.float32],
+            )
+
+            if sklearn_check_version("1.2"):
+                self._check_params_vs_input(X)
```

Comment:
How does ```_check_params_vs_input``` and ```_check_params``` worth with ```_validate_params```?  I also see something similar defined as a method in the onedal object.

Reply:
`_validate_params` make use of the dictionary parameter_constraints to make sure that the input type (array-like, None, "auto" etc) is consistent with expected. On the other hand `_check_params_vs_input` is algorithm specific and allows some specific checks like number of observations must be greater than number of clusters. `_check_params` was replaced by `_check_params_vs_input` in 1.2. The sklearn `_check_params_vs_input` is similar to the onedal check. Let me see if it can be done only once either in sklearnex or onedal.